### PR TITLE
Constrain agent names

### LINKED ISSUE #2172: constrain agent name to only string values
Agent name should only be a string.

Now it is possible, for example, to pass an entity as an agent name, for which no warnings/errors are given.

### PR DESCRIPTION
**Pull request by andraskvr**
# Description

closes #2172 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/resources.py [hunk 1]
```diff
@@ -72,6 +72,8 @@ class resource(object):  # noqa: N801
     _resources: Dict[str, Tuple[Type["Resource"], Dict[str, str]]] = {}
 
     def __init__(self, name: str, id_attribute: str, agent: str):
+        if type(agent) != str:
```
Comment (wouterdb) [inline src/inmanta/resources.py:75]: why not isinstance ?

File: src/inmanta/resources.py [hunk 2]
```diff
@@ -269,11 +271,17 @@ def object_to_id(cls, model_object: DynamicProxy, entity_name: str, attribute_na
                     agent_value = agent_value[0]
 
                 agent_value = getattr(agent_value, el)
+                if type(agent_value) != str and type(agent_value) != DynamicProxy:
```
Comment (wouterdb) [inline src/inmanta/resources.py:274]: Why allow it to be DynamicProxy? I think the whole point was to prevent that?

File: tests/test_resource.py [hunk 3]
```diff
@@ -229,3 +230,44 @@ class MyResource(inmanta.resources.Resource):
...
+        @resource("__config__::XResource", agent=42, id_attribute="key")
+        class MyResource(inmanta.resources.Resource):
+            """
+                A file on a filesystem
+            """
+
+            fields = ("key", "value", "agent")
+
+
+def test_resource_invalid_agent_name_attribute_type(snippetcompiler):
```
Comment (wouterdb) [inline tests/test_resource.py:250]: Also test passing an entity as agent id (should not be allowed)

File: src/inmanta/resources.py (hunk 2 above)
Comment (arnaudsjs) [inline src/inmanta/resources.py:274]: Shouldn't we put this one after the for-loop and remove the DynamicProxy condition from the if-statement?

Comment (arnaudsjs) [reply src/inmanta/resources.py:274]: This is a duplicate of Wouter his comment.

**Pull request merged**